File: ROS/src/simpleRos/simpleRos/action_client.py
import sys

import rclpy
from action_msgs.msg import GoalStatus
from rclpy.action import ActionClient, GoalResponse
from rclpy.action.client import ClientGoalHandle
from rclpy.node import Node
from rclpy.task import Future
from user_interface.action import Fibonacci
from user_interface.action._fibonacci import Fibonacci_GetResult_Response


class Action_client(Node):
    def __init__(self):
        super().__init__("action_client")
        self.action_client: ActionClient = ActionClient(self, Fibonacci, "fibonacci")
        self.future = Future()
        self.get_result_future = Future()

    def send_goal(self, step: str):
        goal_msg : Fibonacci.Goal = Fibonacci.Goal()
        goal_msg.step = int(step)
        while(not self.action_client.wait_for_server(timeout_sec=1)):
            self.get_logger().info("fibonacci server is not available!!")
        self.future : Future= self.action_client.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        self.future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future : Future):
        goal_handle : ClientGoalHandle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info("Goal rejected!!!")
            return
        self.get_logger().info("Goal Accepted!!!")
        self.get_result_future : Future = goal_handle.get_result_async()
        self.get_result_future.add_done_callback(self.get_result_callback)
        self.get_logger().info("end response callback!!!")

    def get_result_callback(self, future : Future):
        result : Fibonacci_GetResult_Response = future.result()
        if result.status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info(f"result: {result.result.seq}")
        if result.status == GoalStatus.STATUS_ABORTED:
            self.get_logger().info("aborted!!")

    def feedback_callback(self, msg):
        feedback : Fibonacci.Feedback = msg.feedback
        self.get_logger().info( f"Received feedback : {feedback.temp_seq}")

def main(args = None):
    rclpy.init(args=args)
    node = Action_client()
    node.send_goal(sys.argv[1])
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()

if __name__ == "__main__":
    main()

File: ROS/src/simple_ros_cpp/src/action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "user_interface/action/fibonacci.hpp"
#include <chrono>

using namespace std;
using namespace std::placeholders;
using namespace std::chrono_literals;

class ActionClient : public rclcpp::Node
{
public:
    ActionClient()
        : Node("action_client")
    {
        _action_client = rclcpp_action::create_client<Fibonacci>(this, "fibonacci");
    }

    void send_goal(char step[])
    {
        while (!_action_client->wait_for_action_server(1s))
        {
            RCLCPP_INFO(get_logger(), "server is not available!!!");
        }
        auto goal_msg = Fibonacci::Goal();
        goal_msg.step = stoi(step);
        auto options = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
        options.goal_response_callback = std::bind(&ActionClient::goal_response_callback, this, _1);
        options.feedback_callback = std::bind(&ActionClient::feedback_callback, this, _1, _2);
        options.result_callback = std::bind(&ActionClient::result_callback, this, _1);
        _action_client->async_send_goal(goal_msg, options);
    }

private:
    typedef user_interface::action::Fibonacci Fibonacci;
    typedef rclcpp_action::ClientGoalHandle<Fibonacci> GoalHandleFibonacci;
    rclcpp_action::Client<Fibonacci>::SharedPtr _action_client;
    void goal_response_callback(const GoalHandleFibonacci::SharedPtr &goal_handle)
    {
        if (goal_handle)
        {
            RCLCPP_INFO(get_logger(), "Goal accepted by server!");
        }
    }
    void feedback_callback(GoalHandleFibonacci::SharedPtr, const std::shared_ptr<const Fibonacci::Feedback> feedback)
    {
        RCLCPP_INFO(get_logger(), "feedback :");
        for (auto seq : feedback->temp_seq)
        {
            RCLCPP_INFO(get_logger(), "%ld", seq);
        }
    }
    void result_callback(const GoalHandleFibonacci::WrappedResult &result)
    {
        switch (result.code)
        {
        case rclcpp_action::ResultCode::SUCCEEDED:
            RCLCPP_INFO(get_logger(), "result :");
            for (auto seq : result.result->seq)
            {
                RCLCPP_INFO(get_logger(), "%ld", seq);
            }
            return;
        case rclcpp_action::ResultCode::CANCELED:
            RCLCPP_INFO(get_logger(), "canceled!!");
            return;
        case rclcpp_action::ResultCode::ABORTED:
            RCLCPP_INFO(get_logger(), "aborted!!");
            return;
        case rclcpp_action::ResultCode::UNKNOWN:
            RCLCPP_INFO(get_logger(), "unknown!!");
            return;
        }
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ActionClient>();
    node->send_goal(argv[1]);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
